ext/drpttracer/drpttracer.c

```diff
@@ -0,0 +1,597 @@
+#include <stdint.h>
+#include <linux/perf_event.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/ioctl.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#include "../../core/unix/include/syscall_linux_x86.h" // for SYS_perf_event_open
+#include "dr_api.h"
+#include "drpttracer.h"
+
+#if !defined(X86_64) || !defined(LINUX)
+#    error "This is only for Linux x86_64."
+#endif
+
+#ifdef DEBUG
+#    define ASSERT(x, msg) DR_ASSERT_MSG(x, msg)
+#else
+#    define ASSERT(x, msg) /* nothing */
+#endif
+
+#ifdef DEBUG
+#    define ERRMSG(...)                      \
+        do {                                 \
+            dr_fprintf(STDERR, __VA_ARGS__); \
+        } while (0)
+#else
+#    define ERRMSG(...) /* nothing */
+#endif
+
+/* drpttracer supports multiple types of tracing. However, perf_event_attr's
+ * initialization causes significant overhead. So drpttracer stores all tracing types'
+ * perf_event_attr as global static variables, then drpttracer can directly use them when
+ * the caller want to start tracing.
+ */
+static struct perf_event_attr user_only_pe_attr;
+static struct perf_event_attr kernel_only_pe_attr;
+static struct perf_event_attr user_kernel_pe_attr;
+
+/* pt_shared_metadata is used to store shared meta data: cpu_family, cpu_model,
+ * cpu_stepping.
+ */
+static pt_metadata_t pt_shared_metadata;
+
+/***************************************************************************
+ * UTTILITY FUNCTIONS
+ */
+
+static bool
+read_file_to_buf(IN const char *filename, OUT char **buf, OUT uint64_t *buf_size)
+{
+    char *local_buf = NULL;
+    uint64_t local_buf_size = 0;
+    file_t f = dr_open_file(filename, DR_FILE_READ);
+    if (f == INVALID_FILE) {
+        ERRMSG("failed to open file %s\n", filename);
+        return false;
+    }
+
+    if (!dr_file_size(f, &local_buf_size)) {
+        ERRMSG("failed to get file size of %s\n", filename);
+        dr_close_file(f);
+        return false;
+    }
+
+    local_buf = (char *)dr_global_alloc(local_buf_size);
+    if (!dr_read_file(f, local_buf, local_buf_size)) {
+        ERRMSG("failed to read file %s\n", filename);
+        dr_global_free(local_buf, local_buf_size);
+        dr_close_file(f);
+        return false;
+    }
+
+    *buf = local_buf;
+    *buf_size = local_buf_size;
+
+    dr_close_file(f);
+    return true;
+}
+
+/*
+ * This function is used to copy data from ring buffer to a output buffer.
+ * ring buffer:
+ *  if head > tail:
+ *   vaild data = [tail, head]
+ *  else if head < tail:
+ *   vaild data = [tail, ring_size) + [0, head]
+ *  else:
+ *   vaild data = empty
+ */
+static void
+read_ring_buf_to_buf(IN void *drcontext, IN uint8_t *base, IN uint64_t base_size,
+                     IN uint64_t head, IN uint64_t tail, OUT void **data_buf,
+                     OUT uint64_t *data_buf_size)
+{
+    if (base == NULL) {
+        ERRMSG("try to read from NULL buffer\n");
+        return;
+    }
+    uint64_t head_offs = head % base_size;
```

Comment:
As discussed offline, let's add an assert to check that there's no overwritten data. Also add description about the case where overwriting the data causes issues in decode (partially written records case you described offline), so we need to make sure that the buffer is large enough.

Reply:
So we are seeing overflows -- in what % of syscalls?  What was the default buffer size again, and how many branches does it hold?

Reply:
Could we need to add the ring buffer's details in our code?  We didn't handle these cases now. How about we put it in our design doc?

Reply:
I don't think we're seeing overflows currently. It's just that we understand the implications of an overflow better now. My suggestion was to clearly document something like the following.

```
The user must ensure that the buffer is sufficiently large. Overflows in the ring buffer will overwrite old data, which may cause issues during decoding, e.g. pt2ir may be unable to decode partial PT traces, or some entry may be partially overwritten (ideally, we should read backwards).
```

Reply:
If the user must ensure the buffer is large enough, are we now exposing the buffer size to the user?  That was not the case when I first reviewed the code.

Reply:
> So we are seeing overflows -- in what % of syscalls? What was the default buffer size again, and how many branches does it hold?

Currently, we don't set the default buffer size and the caller needs to set the buffer size. And if the buffer is overflow, the end trace func will return an error status.

Reply:
> If the user must ensure the buffer is large enough, are we now exposing the buffer size to the user? That was not the case when I first reviewed the code.

Yes. I modified the code. Now the API exposing the buffer size to the user.

Reply:
Perf sets the PT's buffer size to 4MiB. We can set the default size as 4MiB too.
https://github.com/torvalds/linux/blob/6e7765cb477a9753670d4351d14de93f1e9dbbd4/tools/perf/Documentation/perf-intel-pt.txt#L676-L684